simulator.py

```diff
@@ -0,0 +1,114 @@
+"""
+Create a simulator to calculate pnl
+"""
+
+import copy
+import uuid
+from thousandaire.dataset import Data
+
+DATASET = 'price_dataset'
+
+def decode_data(data):
+    """
+    Decode while in multi-process.
+
+    In multi-process, simulator will get the data reference
+    which alpha need. To prevent polluteing the origin data, we need to
+    copy a local version.
+    Since 'target_data' may also in 'others', that we should deepcopy the data
+    separately not simply using copy.deepcopy(data).
+    """
+    return {
+        'workdays' : copy.deepcopy(data['workdays']),
+        'target_data' : copy.deepcopy(data['target_data']),
+        'others' : copy.deepcopy(data['others'])}
+
+class Simulator:
+    """
+    Simulate the alpha.
+    """
+    def __init__(self, settings, data, cost_object):
+        self.cost_object = cost_object
+        self.data = decode_data(data)
+        self.settings = settings
+        self.__result = Data(
+            'result', ['pnl', 'cost', 'positions_raw', 'positions_np'])
+        self.__portfolio = list()
+        self.__key = uuid.uuid4()
+        self.initialize_data()
+
+    def next_date(self):
+        """
+        Move the simulating date.
+        """
+        for dataset in self.data['others'].values():
+            dataset.next_date(self.__key)
+        self.data['target_data'].next_date(self.__key)
+        self.data['workdays'].next_date(auth_key=self.__key)
+
+    def produce_pnl(self):
+        """
+        Calculate pnl for alphas.
+        This method will be called by simulation day by day while inputting
+        a new portfolio.
+
+        While calculating pnl,
+        we should take a look at our asset and trading cost.
+            (1) Total asset
+                We assume that our assets are the same everyday (equal to
+                self.initial_asset). To achieve this goal, we open a virtual
+                account to borrow/lend money to the investing account (this
+                account is self.cur_asset[self.base]).
+            (2) Trading cost
+                There are different functions about calculating trading cost
+                in different investing targets.
+                This method should import different modules while targets are
+                different.
+
+                In our template function, we calculate trading cost while
+                selling out the asset.
+                The trading_cost function could be built by alpha writers.
+                The API will adjust in the future, while the fuction needs
+                more data.
+        """
+        self.cost_object.reset()
+        self.cost_object(self.__portfolio, self.data['target_data'])
+        profit, cost = self.cost_object.result()
```

Comment:
Why should not let it be `profit, cost = self.cost_function(self.__portfolio, self.data['target_data'])`?